### DISCUSSION TITLE
I want to know how to transfer a point in the world coordinate system to screen coordinates. I wrote an API myself and found that the x-axis coordinates are always several tens of pixels apart. I don't know why, the y-screen coordinates can be aligned, and the depth cache is also acceptable

### DISCUSSION THREAD

**Discussion by Haidiye00** [google/filament#6906] [General]
I want to know how to transfer a point in the world coordinate system to screen coordinates. I wrote an API myself and found that the x-axis coordinates are always several tens of pixels apart. I don't know why, the y-screen coordinates can be aligned, and the depth cache is also acceptable
![9MWR8BWJdQ](https://github.com/google/filament/assets/132528822/cd32aeb1-ea3d-4832-98dd-20d753b512ec)
this is test result:
![NfCCRg1EmR](https://github.com/google/filament/assets/132528822/c9529adf-d281-4621-8e10-13e7dfae2c3a)
Actually, my fundamental requirement is **how to convert screen coordinates into world coordinate system coordinates**. That is to implement a API named: **mouse_ To_ 3D** 

Forgive me for not being able to copy the code because it cannot be copied, you understand。
Thanks

**Comment by jeanlemotan** [comment]
Each pixel in 2D corresponds with a line in 3D that connects a point in the near plane to a point in the far plane, so to get a 3D point from a 2D pixel, you'll need a depth value as well.
In general, this is how the code looks like:

```
bool Unproject(const math::mat4f& inverseViewProjection, float clipSpaceX, float clipSpaceY, float clipSpaceZ, math::vec3f& out)
{
	math::vec4f tmp = math::transform(inverseViewProjection, math::vec4f(clipSpaceX, clipSpaceY, clipSpaceZ, 1.f));
	if (std::abs(tmp.w) > std::numeric_limits<float>::epsilon())
	{
		out = math::vec3f(tmp.x, tmp.y, tmp.z) / tmp.w;
		return true;
	}
	return false;
}
```

You give it the inverse view projection matrix and the 2D point in clip space (-1..1 coordinates for both X, Y and Z (depth)). It will calculate the 3D coords in the `out` argument.
